text/3371-cargo-target-dir-templates.md

```diff
@@ -0,0 +1,424 @@
+1. Allows using a different disk, partition or mount point for cargo artifacts
+1. Avoids having to set `CARGO_TARGET_DIR` for every project to get the same effect as proposed here
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+<!--
+Explain the proposal as if it was already included in the language and you were teaching it to another Rust programmer. That generally means:
+
+- Introducing new named concepts.
+- Explaining the feature largely in terms of examples.
+- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
+- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
+- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?
+
+For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+-->
+
+For a single project, it is possible to use the `CARGO_TARGET_DIR` environment variable (or the `target-dir` TOML config option or the `--target-dir` command-line flag) with either an absolute or relative path to change the position of the `target/` directory used for build artifacts during compilation with Cargo.
+
+While this option is useful for single-project environments (simple CI builds, builds through other build systems like Meson or Bazel), in multi-projects environment, like personal machines or repos with multiple workspaces, it conflates every build directory under the configured path: `CARGO_TARGET_DIR` directly replaces the `<workspace>/target/` directory.
+
+Templating introduces one new templating key for `CARGO_TARGET_DIR`, in the same spirit as [the index configuration format][icf]:
+
+- `{manifest-path-hash}`: a hash of the manifest's absolute path as a path. This is **not** an absolute path.
+
+It can be used like this: `CARGO_TARGET_DIR="$HOME/.cache/cargo-target-dirs/{manifest-path-hash}"`.
+
+When compiling `/home/ferris/src/cargo/` with user `ferris`, `manifest-path-hash` would be something like `ab/cd/<rest of hash>` and the artifacts would be found in `/home/ferris/.cache/cargo-target-dirs/ab/cd/<rest of hash>/...`. Note the hash used and the path derived from that for `{manifest-path-hash}` are implementation details and the values here are just example.
```

Comment:
```suggestion
When compiling `/home/ferris/src/cargo/` with user `ferris`, `manifest-path-hash` would be something like `ab/cd/<rest of hash>` and the artifacts would be found in `/home/ferris/.cache/cargo-target-dirs/ab/cd/<rest of hash>/...`. Note the hash used and the path derived from that for `{manifest-path-hash}` are implementation details and the values here are just an example.
```

Reply:
Done in 0a38c0767eeae0d6f0bdb433c9207c167e4f3245